ReactiveCocoa/Swift/SignalProducer.swift

```diff
@@ -176,31 +164,36 @@ public struct SignalProducer<Value, Error: ErrorType> {
 			}
 		}
 
+		let pushLock = NSLock()
+		pushLock.name = "org.reactivecocoa.ReactiveCocoa.SignalProducer.buffer.push"
 		let bufferingObserver: Signal<Value, Error>.Observer = Observer { event in
-			// Send serially with respect to other senders, and never while
-			// another thread is in the process of replaying.
-			dispatch_sync(queue) {
-				let originalState = state.modify { state in
-					var mutableState = state
-
-					if let value = event.value {
-						mutableState.addValue(value, upToCapacity: capacity)
-					} else {
-						// Disconnect all observers and prevent future
-						// attachments.
-						mutableState.terminationEvent = event
-						mutableState.observers = nil
-					}
-
-					return mutableState
+			// Serialize pushing the events with a lock to prevent
+			// different observers from receiving the events in different order
+			// in case multiple threads push events at the same time.
+			pushLock.lock()
```

Comment:
What's unfair about NSLock locking? Sending is still serial, but replaying threads are protected by per-observer locks, not this lock. There are comments describing it too.